retrieval/targetmanager.go

```diff
@@ -417,79 +418,102 @@ func providersFromConfig(cfg *config.ScrapeConfig) map[string]TargetProvider {
-				model.JobLabel:         model.LabelValue(cfg.JobName),
-			},
-		}
-		for _, lset := range labelsets {
-			for ln, lv := range lset {
-				if _, ok := labels[ln]; !ok {
-					labels[ln] = lv
-				}
-			}
+// populateLabels builds a label set from the given label set and scrape configuration.
+// It returns a label set before relabeling was applied as the second return value.
+// Returns a nil label set if the target is dropped during relabeling.
+func populateLabels(lset model.LabelSet, cfg *config.ScrapeConfig) (res, orig model.LabelSet, err error) {
+	if _, ok := lset[model.AddressLabel]; !ok {
+		return nil, nil, fmt.Errorf("no address")
+	}
+	// Copy labels into the labelset for the target if they are not
+	// set already. Apply the labelsets in order of decreasing precedence.
+	scrapeLabels := model.LabelSet{
+		model.SchemeLabel:      model.LabelValue(cfg.Scheme),
+		model.MetricsPathLabel: model.LabelValue(cfg.MetricsPath),
+		model.JobLabel:         model.LabelValue(cfg.JobName),
+	}
+	for ln, lv := range scrapeLabels {
+		if _, ok := lset[ln]; !ok {
+			lset[ln] = lv
 		}
-
-		if _, ok := labels[model.AddressLabel]; !ok {
-			return nil, fmt.Errorf("instance %d in target group %s has no address", i, tg)
+	}
+	// Encode scrape query parameters as labels.
+	for k, v := range cfg.Params {
+		if len(v) > 0 {
+			lset[model.LabelName(model.ParamLabelPrefix+k)] = model.LabelValue(v[0])
 		}
+	}
 
-		preRelabelLabels := labels
+	preRelabelLabels := lset
+	lset = relabel.Process(lset, cfg.RelabelConfigs...)
 
-		labels := relabel.Process(labels, cfg.RelabelConfigs...)
+	// Check if the target was dropped.
+	if lset == nil {
+		return nil, nil, nil
+	}
 
-		// Check if the target was dropped.
-		if labels == nil {
-			continue
+	// addPort checks whether we should add a default port to the address.
+	// If the address is not valid, we don't append a port either.
+	addPort := func(s string) bool {
+		// If we can split, a port exists and we don't have to add one.
+		if _, _, err := net.SplitHostPort(s); err == nil {
+			return false
 		}
-		// If no port was provided, infer it based on the used scheme.
-		addr := string(labels[model.AddressLabel])
-		if !strings.Contains(addr, ":") {
-			switch labels[model.SchemeLabel] {
-			case "http", "":
-				addr = fmt.Sprintf("%s:80", addr)
-			case "https":
-				addr = fmt.Sprintf("%s:443", addr)
-			default:
-				return nil, fmt.Errorf("invalid scheme: %q", cfg.Scheme)
-			}
-			labels[model.AddressLabel] = model.LabelValue(addr)
+		// If adding a port makes it valid, the previous error
+		// was not due to an invalid address and we can append a port.
+		_, _, err := net.SplitHostPort(s + ":1234")
+		return err == nil
+	}
+	// If it's an address with no trailing port, infer it based on the used scheme.
+	if addr := string(lset[model.AddressLabel]); addPort(addr) {
+		switch lset[model.SchemeLabel] {
+		case "http", "":
+			addr = fmt.Sprintf("%s:80", addr)
```

Comment:
Ah I think I get it. Incoming addrs are already correctly formatted (with `[]` for IPv6), and `net.JoinHostPort()` actually adds another layer of `[]` around the address part if the brackets are already present. Would be good to make clear (in a comment?) that any addrs that are inputs here are assumed to not be a raw IPv6 address (`::1`), anymore but already wrapped. From something thinking that ends up being clear anyways because the typical addr in Go can always contain a port, and then the bracket notation is necessary in any case for disambiguation. Hmm...